sqlite.go

```diff
@@ -0,0 +1,542 @@
+// For details see https://sqlite.org/c3ref/open.html#urifilenames.
+//
+//
+// Memory Mode
+//
+// In-memory databases are popular for tests.
+// If opening a database in memory mode with the "?mode=memory" query
+// parameter, make sure immediatley after opening to set
+//
+//	db.SetMaxOpenConns(1)
+//
+// This driver does not support shared cache mode, as the SQLite
+// authors recommend against it:
+//
+//	"Any use of shared cache is discouraged."
+//	https://www.sqlite.org/c3ref/enable_shared_cache.html
+//
+// For far more realistic in-memory database testing, use
+//
+//	PRAGMA journal_mode=WAL;
+//	PRAGMA synchronous=OFF;
+//
+// On any modern OS, this will generate syscalls as part of writing
+// data, but will never block on I/O.
+//
+//
+// Binding Types
+//
+// SQLite is flexible about type conversions, and so is this driver.
+// Almost all "basic" Go types (int, float64, string) are accepted and
+// directly mapped into SQLite, even if they are named Go types.
+// The time.Time type is also accepted (described below).
+// Unrecognized types are tested to be implementations of
+// encoding.TextMarshaler or json.Marshaler, if they are, those are
+// called and the byte output is stored in SQLite.
+//
+//
+// Binding Time
+//
+// While SQLite3 has no strict time datatype, it does have a series of built-in
+// functions that operate on timestamps that expect columns to be in one of many
+// formats: https://sqlite.org/lang_datefunc.html
+//
+// When encoding a time.Time into one of SQLite's preferred formats, we use the
+// shortest timestamp that the time.Time can be represented using, one of:
+//
+//	2. YYYY-MM-DD HH:MM
+//	3. YYYY-MM-DD HH:MM:SS
+//	4. YYYY-MM-DD HH:MM:SS.SSS
+//
+// If the time.Time is not UTC (strongly consider storing times in UTC!),
+// we follow SQLite's norm of appending "[+-]HH:MM" to the above formats.
+//
+// It is common for efficiency to store a time as seconds-since-epoch in an
+// INTEGER column. If you want to do that, pass the result of time.Time.Unix
+// to the driver.
+//
+//
+// Reading Time
+//
+// In general, time is hard to extract from SQLite as a time.Time.
+// If a column is defined as DATE or DATETIME, then the text can be parsed
+// as TimeFormat and returned as a time.Time.
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"encoding"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/tailscale/sqlite/cgosqlite"
+	"github.com/tailscale/sqlite/sqliteh"
+)
+
+// TimeFormat is the string format this driver uses to store
+// microsecond-precision time in SQLite in text format.
+const TimeFormat = "2006-01-02 15:04:05.000-0700"
+
+func init() {
+	sql.Register("sqlite3", drv)
+}
+
+// ErrMsg sets the driver to collect the sqlite3_errmsg string on error.
+// This is useful but generates extra string allocations.
+// Turn it off if you have written a program with an SQL error on a hot path.
+var ErrMsg = true
```

Comment:
Can you make this an open option on the DB URL instead? It feels dirty to have mutable package-level variables.

Reply:
I haven't got into the business of parsing the URL yet and I would like to avoid it. I'll investigate an API for this later. For now I've removed it.